pymc_marketing/prior.py

```diff
@@ -0,0 +1,893 @@
+        """Create a Prior from the dictionary format.
+
+        Parameters
+        ----------
+        json : dict[str, Any]
+            The dictionary format of the prior.
+
+        Returns
+        -------
+        Prior
+            The prior distribution.
+
+        Examples
+        --------
+        Convert prior in the dictionary format to a Prior instance.
+
+        .. code-block:: python
+
+            from pymc_marketing.prior import Prior
+
+            json = {
+                "dist": "Normal",
+                "kwargs": {"mu": 0, "sigma": 1},
+            }
+
+            dist = Prior.from_json(json)
+            dist
+            # Prior("Normal", mu=0, sigma=1)
+
+        """
+        if not isinstance(json, dict):
+            msg = (
+                "Must be a dictionary representation of a prior distribution. "
+                f"Not of type: {type(json)}"
+            )
+            raise ValueError(msg)
+
+        dist = json["dist"]
+        kwargs = json.get("kwargs", {})
+
+        def handle_value(value):
+            if isinstance(value, dict):
+                return cls.from_json(value)
+
+            if isinstance(value, list):
+                return np.array(value)
+
+            return value
+
+        kwargs = {param: handle_value(value) for param, value in kwargs.items()}
+        centered = json.get("centered", True)
+        dims = json.get("dims")
+        if isinstance(dims, list):
+            dims = tuple(dims)
+        transform = json.get("transform")
+
+        return cls(dist, dims=dims, centered=centered, transform=transform, **kwargs)
+
+    def constrain(self, lower: float, upper: float, **kwargs) -> Prior:
+        """Create a new prior that is constrained to the given bounds.
+
+        Wrapper around `pm.find_constrained_prior` where the initial guess is
+        the current parameters. Will error out if no parameters are given.
+
+        Parameters
+        ----------
+        lower : float
+            The lower bound.
+        upper : float
+            The upper bound.
+        kwargs : dict
+            Additional arguments to pass to `pm.find_constrained_prior`.
+
+        Returns
+        -------
+        Prior
+            The new prior that is constrained to domain.
+
+        Examples
+        --------
+        Create a Beta distribution that is constrained between 0.5 and 0.8
+        using initial parameter values of alpha=2 and beta=1 for the
+        optimization.
+
+        .. code-block:: python
+
+            dist = Prior(
+                "Beta",
+                alpha=2,
+                beta=1,
+            ).constrain(lower=0.5, upper=0.8)
+
+        """
+
+        if self.transform:
+            raise ValueError("Can't constrain a transformed variable")
+
+        new_parameters = pm.find_constrained_prior(
+            self.pymc_distribution,
+            lower=lower,
+            upper=upper,
+            init_guess=self.parameters,
+            **kwargs,
+        )
```

Comment:
We need to handle the exception when this `find_constrained_prior` fails. Maybe add a reasonable fallback mechanism? Or just handle the exception and inform the user with an informative error message?